IdentityCore/src/requests/broker/MSIDGCDStarvationDetector.m

```diff
@@ -0,0 +1,127 @@
+// THE SOFTWARE.
+
+#import "MSIDGCDStarvationDetector.h"
+#import "MSIDLogger+Internal.h"
+
+static NSTimeInterval starvationCheckTimeout = 0.01; //10 ms - timeout for detecting if GCD thread pool is starved
+static NSTimeInterval starvationCheckInterval = 0.1; //100 ms - interval between starvation checks
+static NSTimeInterval maxMonitoringDuration = 15.0; //15 seconds - maximum monitoring duration to prevent indefinite running
+
+@interface MSIDGCDStarvationDetector()
+
+@property (nonatomic) BOOL running;
+@property (nonatomic) BOOL shouldStop;
+@property (nonatomic) NSThread *monitorThread;
+@property (nonatomic) NSTimeInterval gcdStarvedDuration;
+@property (nonatomic) NSInteger totalPingCount;
+@property (nonatomic) NSInteger starvedPingCount;
+@property (nonatomic) NSDate *monitoringStartTime;
+
+@end
+
+@implementation MSIDGCDStarvationDetector
+
+#pragma mark - Public API
+
+// We need to start the monitoring in a customized thread to avoid the current thread which invokes this detector being blocked
+- (void)startMonitoring
+{
+    @synchronized(self) {
+        if (self.running) return;
+
+        self.shouldStop = NO;
+        self.monitoringStartTime = [NSDate date];
+        _monitorThread = [[NSThread alloc] initWithTarget:self
+                                                 selector:@selector(monitorLoop)
+                                                   object:nil];
+        _monitorThread.name = [NSString stringWithFormat:@"%@-%@", @"com.microsoft.msid.MSIDGCDStarvationDetector.thread", [NSUUID UUID].UUIDString];
+        [_monitorThread start];
+        self.running = YES;
+    }
+}
+
+- (NSTimeInterval)stopMonitoring {
+    @synchronized (self) {
+        self.shouldStop = YES;
+        if (self.monitorThread) {
+            [self.monitorThread cancel];
+        }
+        self.monitorThread = nil;
+        self.running = NO;
+        return self.gcdStarvedDuration;
+    }
+}
+
+- (void)monitorLoop {
+    @autoreleasepool {
+        MSID_LOG_WITH_CTX(MSIDLogLevelInfo, nil, @"GCDStarvationDetector -- started on thread: %@", [NSThread currentThread]);
+        
+        while (YES) {
+            if (self.shouldStop || [NSThread currentThread].isCancelled) {
+                break;
+            }
+
+            NSTimeInterval elapsed = [[NSDate date] timeIntervalSinceDate:self.monitoringStartTime];
+            if (elapsed >= maxMonitoringDuration) {
+                MSID_LOG_WITH_CTX(MSIDLogLevelWarning, nil, @"GCDStarvationDetector -- reached maximum duration (%.2fs), stopping", elapsed);
+                [self stopMonitoring];
+                break;
+            }
+
+            @synchronized (self) {
+                BOOL starved = [self isThreadStarvedWithTimeout:starvationCheckTimeout];
+                self.totalPingCount += 1;
+                if (starved) {
+                    self.gcdStarvedDuration += (starvationCheckTimeout + starvationCheckInterval);
```

Comment:
When the code enters the while(YES) loop for the first time, there is no sleep happened before GCD operation, so we don't count `starvationCheckInterval` for the first time?